abel/basex.py

```diff
@@ -110,280 +120,289 @@ def basex_transform(data, nbf='auto', basis_dir='./', dr=1.0, verbose=True,
-        2D arrays given by the basis set calculation function
+    rawdata : m x n numpy array
+        the raw image. This is the right half (with the axis) of the image.
+    Ai : n x n numpy array
+        2D array given by the transform-calculation function
     dr : float
         pixel size. This only affects the absolute scaling of the output.
 
 
     Returns
     -------
-    IM : NxM numpy array
+    IM : m x n numpy array
         The abel-transformed image, a slice of the 3D distribution
     """
 
-    # Reconstructing image  - This is where the magic happens
-    Ci = scipy.dot(scipy.dot(vert_left, rawdata), horz_right) # previously: vert_left.dot(rawdata).dot(horz_right)
+    # Note that our images are stored with matrix rows corresponding to
+    # horizontal image lines. This is consistent with the Matlab and C++
+    # BASEX implementations, but is the opposite to the article notation.
+    # Thus all matrices are transposed and multiplied backwards with
+    # respect to the article.
+    # The vertical transform is not applied, since without regularization
+    # its overall effect is an identity transform.
 
-    # use an heuristic scaling factor to match the analytical abel transform
-    # For more info see https://github.com/PyAbel/PyAbel/issues/4
-    MAGIC_NUMBER = 1.1122244156826457
-    Ci *= MAGIC_NUMBER/dr
-    IM = scipy.dot(scipy.dot(Mc_vert, Ci), Mc_horz.T)    # Previously: Mc_vert.dot(Ci).dot(Mc_horz.T)
-    # P = dot(dot(Mc,Ci),M.T) # This calculates the projection,
-    # which should recreate the original image
+    # Reconstructing image  - This is where the magic happens
+    IM = scipy.dot(rawdata, Ai) / dr
+    # P = dot(dot(Mc,Ci),M.T) # This calculates the projection, !! not
+    # which should recreate the original image                  !! really
     return IM
 
 
-def _get_left_right_matrices(M_vert, M_horz, Mc_vert, Mc_horz):
+def _get_Ai(M, Mc, reg):
     """ An internal helper function for no-up/down-asymmetry BASEX:
-        given basis sets  M_vert, M_horz, Mc_vert, Mc_horz,
-        return M_left and M_right matrices
+        given basis sets M, Mc,
+        return matrix of inverse Abel transform
     """
 
-    nbf_vert, nbf_horz = np.shape(M_vert)[1], np.shape(M_horz)[1]
-    q_vert, q_horz = 0, 0  # No Tikhonov regularization
-    E_vert, E_horz = np.identity(nbf_vert)*q_vert, np.identity(nbf_horz)*q_horz
+    # Ai  is the overall (horizontal) transform matrix,
+    #     corresponds to A^T Z in the article
+    # reg corresponds to q_1^2 in the article
 
-    vert_left  = scipy.dot( inv(scipy.dot(Mc_vert.T, Mc_vert) + E_vert),  Mc_vert.T)
-    horz_right = scipy.dot( M_horz, inv(scipy.dot(M_horz.T, M_horz) + E_horz) )
-    
-    # previously: 
-    # vert_left = inv(Mc_vert.T.dot(Mc_vert) + E_vert).dot(Mc_vert.T)
-    # horz_right = M_horz.dot(inv(M_horz.T.dot(M_horz) + E_horz))
-    
+    if reg == 0.0:
+        Ai = scipy.dot(inv(M.T), Mc.T)  # (this works only for nbf == n)
```

Comment:
See https://github.com/PyAbel/PyAbel/commit/34f6fb620975a343c52dd6fbcad4eefe5c6dffee.
@DanHickstein, is it important?

Reply:
Not too important, and for linalg there may indeed be a difference but scipy does not implement the dot product, [it's a numpy function](https://github.com/scipy/scipy/blob/016a6ef37f68ebdb9a2a0901f2a1131562441acc/scipy/__init__.py#L71),
```py
>>> id(numpy.dot) == id(scipy.dot)
True
```
so there is also no real reason to import it from scipy.

Reply:
Seems to be true. At least [scipy.linalg docs](https://docs.scipy.org/doc/scipy/reference/tutorial/linalg.html) use `A.dot(B)` instead of `scipy.dot(A, B)`. Let's see what @DanHickstein can say.

If we are going to convert `.dot` back `ndarray`'s method, do you think whether
`R = M.dot(inv((M.T).dot(M) + E))`
is more readable than
`R = M.dot(inv(M.T.dot(M) + E))`
or not?

(Conversion to `numpy.matmul` or `@`, as [`numpy.dot` description](https://docs.scipy.org/doc/numpy/reference/generated/numpy.dot.html) suggests is probably not a good option yet.)

Reply:
At one point I was switching all of the linear algebra to scipy, since I thought that it was faster than numpy in the case where the user didn't have fast (Apple Accelerate Framework or Intel MKL) BLAS/LAPACK libraries. I didn't realize that the dot function was actually in numpy. Fine with me to swith it to numpy.

Reply:
I think that @rth is just wanting `scipy.dot` to be changed to `np.dot`, and not necessarily to use the `M.dot(X)` syntax. I think that both are equally clear, but that it does make some sense to use `np.dot` if `scipy.dot` simply redirects to `np.dot` anyway.